### PR TITLE
Disconnected environment using a URL to populate the repository

### PR DESCRIPTION
**Pull request by cooktheryan**
Signed-off-by: Ryan Cook <rcook@redhat.com>

### CODE REVIEW

**Comment by cooktheryan** [comment]
Need to still create the disconnected with localPath test and need to write tests for the full-suite to verify all things are being placed

**Comment by cooktheryan** [comment]
@josephsawaya @sallyom should I remove localpath? I will still use part of the code for copying and such but like if it's on the system already why would I need to load it 🤦‍♂️

**Comment by josephsawaya** [comment]
> @josephsawaya @sallyom should I remove localpath? I will still use part of the code for copying and such but like if it's on the system already why would I need to load it man_facepalming

Do you mean you want to remove the localPath option from the config?

**Comment by cooktheryan** [comment]
@josephsawaya ya just the localpath option for disconnected. There just isn't a world I can think of where that would be used vs http or usb

File: pkg/engine/disconnected.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+func localDevicePull(name, device, trimDir string, image bool) (id string, err error) {
+	// Need to use the filetransfer method to populate the directory from the localPath
+	ctx := context.Background()
+	conn, err := bindings.NewConnection(ctx, "unix://run/podman/podman.sock")
+	if err != nil {
+		klog.Error("Failed to create connection to podman")
+		return "", err
+	}
+	// Ensure that the device is present
+	_, exitCode, err := localDeviceCheck(name, device, "-"+trimDir)
```
Comment (josephsawaya) [inline pkg/engine/disconnected.go:116]: When you call this in the image method (pkg/engine/image.go:118), you pass "-" + trimDir as an arg, so is prepending "-" also necessary in this call to `localDeviceCheck`?

File: pkg/engine/disconnected.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+		return "", err
+	}
+	if exitCode != 0 {
+		// remove the diff file
+		cache := "/opt/.cache/" + name + "/"
+		dest := cache + "/" + "HEAD"
+		err = os.Remove(dest)
+		klog.Info("Device not present...requeuing")
+		return "", nil
+	} else if exitCode == 0 {
```
Comment (sallyom) [inline pkg/engine/disconnected.go:128]: this else isn't needed - you'd have returned if exitCode != 0

File: pkg/engine/image.go [hunk 3]
```diff
@@ -45,29 +61,117 @@ func (i *Image) Apply(ctx, conn context.Context, currentState, desiredState plum
...
-		klog.Error("Failed getting data from ", i.Url)
-		return err
+		// remove the image if it exists
+		if _, err := os.Stat(pathToLoad); err == nil {
+			klog.Info("Flushing image from device ", pathToLoad)
+			flushImages(pathToLoad)
+		}
+		klog.Info("URL not present...requeuing")
+		return nil
+	} else if data.StatusCode == http.StatusOK {
```
Comment (sallyom) [inline pkg/engine/image.go:76]: can remove the else and indentation here

File: pkg/engine/image.go [hunk 4]
```diff
@@ -45,29 +61,117 @@ func (i *Image) Apply(ctx, conn context.Context, currentState, desiredState plum
...
-	data, err := http.Get(i.Url)
+func (i *Image) loadHTTPPodman(ctx, conn context.Context, url, target string) error {
+	imageName := (path.Base(url))
+	pathToLoad := "/opt/" + imageName
+	data, err := http.Get(url)
 	if err != nil {
-		klog.Error("Failed getting data from ", i.Url)
-		return err
+		// remove the image if it exists
+		if _, err := os.Stat(pathToLoad); err == nil {
```
Comment (sallyom) [inline pkg/engine/image.go:70]: klog the err above, then, if err != nil here, return this err
then, don't need the err==nil, just the klog& flushImages below

File: pkg/engine/image.go (hunk 3 above)
Comment (sallyom) [reply pkg/engine/image.go:76]: if data.StatusCode == http.StatusOK {
        if _, err := os.Stat....

**Pull request merged**